src/ms3/score.py

```diff
@@ -1229,111 +1253,437 @@ def store_phrase_excerpts(self, directory: Optional[str] = None):
+
+            suffix: Optional[str], optional
+                It is the string "category identifier" of your excerpts. For instance the name of the output files will
+                in general be ``[original_filename]_[suffix]_[start_mc]-[end_mc].mscx``
+
+            random_skip: Optional[bool], optional
+                This boolean value, if True, will make the method randomly skip extracted
+                excerpts and don't generate them. This parameter is set by default to False.
+        """
+
+        expanded = self.expanded(unfold=True)
+        measures = self.measures(unfold=True)
+
+        df = expanded[expanded["label"].str.contains("{|}|\\|")].copy()
+        last_phrase_label = ""
+        rows_to_drop = []
+        for index, row in df.iterrows():
+            if pd.notna(row["phraseend"]):
+                last_phrase_label = row["phraseend"]
+            if (
+                "|" in row["label"]
+                and last_phrase_label == "}"
+                and pd.isna(row["phraseend"])
+            ):
+                rows_to_drop.append(index)
+        df.drop(rows_to_drop, inplace=True)
+
+        available_mcs = [
+            tuple(
+                measures.iloc[
+                    measures.loc[
+                        measures["mn_playthrough"] == df.iloc[i]["mn_playthrough"]
+                    ].index[0]
+                    + 1 : measures.loc[
+                        measures["mn_playthrough"] == df.iloc[i + 1]["mn_playthrough"]
+                    ].index[0]
+                ]["mc"].values
             )
+            for i in range(len(df) - 1)
+        ]
+
+        available_mcs = [x for x in available_mcs if len(x) >= 2]
+        available_mcs = list(set(available_mcs))
+
+        if available_mcs:
+            for included_mcs in available_mcs:
+                for i in range(len(included_mcs) - 1):
+                    if random_skip and np.random.choice([True, False]):
+                        continue
+                    self.store_measures(
+                        included_mcs=(included_mcs[i : i + 2]),
+                        start_mc_onset=None,
+                        end_mc_onset=None,
+                        user_tempo=tempo,
+                        user_beat_unit=beat_unit,
+                        directory=directory,
+                        suffix=suffix,
+                    )
+        else:
+            self.logger.info("No measures to be stored.")
+
+    def store_phrase_endings(
+        self,
+        tempo: Optional[float] = None,
+        beat_unit: Optional[Fraction] = 1,
+        directory: Optional[str] = None,
+        suffix: Optional[str] = "phrase_end",
+        random_skip: Optional[bool] = False,
+    ):
+        """Calls the self.find_phrase_endings() method to find all phrase endings contained in the score, then stores
+        all corresponding excerpts. A phrase ending is specified to finish on a cadence and to start 2 MCs before the
```

Comment:
The (maximal) excerpt length should be a parameter defaulting to 2.

Reply:
That sounds good. What I will do is that I will remove (comment out) the `find_phrase_endings()` method and instead use the renamed `get_phrase_boundaries()` and keep only the `max_excerpt_length` number of measure from those. I think it would be nice to say that if the max number of measures, for a given phrase, exceeds the number of measures included in that specific phrase, then the excerpt will default to extracting the full phrase. Do you think that is a good idea?